lsm.go

```diff
@@ -0,0 +1,352 @@
+// This should only be used during snapshot recovery.
+func (l *LSM) InsertPart(level SentinelType, part *parts.Part) {
+	l.findLevel(level).Prepend(part)
+}
+
+func (l *LSM) String() string {
+	return l.levels.String()
+}
+
+func (l *LSM) Prefixes(ctx context.Context, prefix string) ([]string, error) {
+	return []string{l.prefix}, nil
+}
+
+func (l *LSM) Scan(ctx context.Context, _ string, _ *dynparquet.Schema, filter logicalplan.Expr, tx uint64, callback func(context.Context, any) error) error {
+	l.RLock()
+	defer l.RUnlock()
+
+	booleanFilter, err := BooleanExpr(filter)
+	if err != nil {
+		return fmt.Errorf("boolean expr: %w", err)
+	}
+	var iterError error
+	l.levels.Iterate(func(node *Node) bool {
+		if node.part == nil { // encountered a sentinel node; continue on
+			return true
+		}
+
+		if node.part.TX() > tx { // skip parts that are newer than this transaction
+			return true
+		}
+
+		if r := node.part.Record(); r != nil {
+			r.Retain()
+			if err := callback(ctx, r); err != nil {
+				iterError = err
+				return false
+			}
+			return true
+		}
+
+		buf, err := node.part.AsSerializedBuffer(nil)
+		if err != nil {
+			iterError = err
+			return false
+		}
+
+		for i := 0; i < buf.NumRowGroups(); i++ {
+			rg := buf.DynamicRowGroup(i)
+			mayContainUsefulData, err := booleanFilter.Eval(rg)
+			if err != nil {
+				iterError = err
+				return false
+			}
+
+			if mayContainUsefulData {
+				if err := callback(ctx, rg); err != nil {
+					iterError = err
+					return false
+				}
+			}
+		}
+		return true
+	})
+	return iterError
+}
+
+// TODO: this should be changed to just retain the sentinel nodes in the lsm struct to do an O(1) lookup
+func (l *LSM) findLevel(level SentinelType) *Node {
+	var list *Node
+	l.levels.Iterate(func(node *Node) bool {
+		if node.part == nil && node.sentinel == level {
+			list = node
+			return false
+		}
+		return true
+	})
+
+	return list
+}
+
+// Merge will merge the given level into an arrow record for the next level using the configured Compact function for the given level.
+// If this is the max level of the LSM an external writer must be provided to write the merged part elsewhere.
+func (l *LSM) merge(level SentinelType, externalWriter func([]*parts.Part) (*parts.Part, int64, int64, error)) error {
+	if int(level) > len(l.configs) {
+		return fmt.Errorf("level %d does not exist", level)
+	}
+	if int(level) == len(l.configs)-1 && externalWriter == nil {
+		return fmt.Errorf("cannot merge the last level without an external writer")
+	}
+	l.metrics.Compactions.WithLabelValues(level.String()).Inc()
+
+	mergeList := []*parts.Part{}
+	var next *Node
+	compact := l.findLevel(level)
+	var iterErr error
+	compact.Iterate(func(node *Node) bool {
+		if node.part == nil { // sentinel encountered
+			switch {
+			case node.sentinel == level: // the sentinel for the beginning of the list
+				return true
+			case node.sentinel == level+1:
+				next = node.next.Load() // skip the sentinel to combine the lists
```

Comment:
I'm slightly confused here: does this mean for example if we merge all l0 parts, we're including the next level parts as well?

Reply:
No it just is finding the first node of the next level that it can point to to continue the linked list